File: src/db/repositories/message_repository.py
"""
Repository for Message operations
"""

import json
import uuid

from src.db.base import db
from src.models.entities import Message, MessageRole, MessageType


class MessageRepository:
    """Repository for message database operations"""

    async def create(
        self,
        conversation_id: str,
        role: MessageRole,
        content: str | None,
        message_type: MessageType,
        media_urls: list[str] | None = None,
        audio_url: str | None = None,
        audio_duration_seconds: int | None = None,
        token_count: int | None = None,
        client_message_id: str | None = None,
        status: str = "delivered",
        is_read: bool = False,
    ) -> Message:
        """Create a new message"""
        message_id = str(uuid.uuid4())
        media_urls_json = json.dumps(media_urls or [])

        query = """
            INSERT INTO messages (
                id, conversation_id, role, content, message_type,
                media_urls, audio_url, audio_duration_seconds, token_count,
                client_message_id, status, is_read
            )
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)
        """

        await db.execute(
            query,
            message_id,
            conversation_id,
            role.value,
            content,
            message_type.value,
            media_urls_json,
            audio_url,
            audio_duration_seconds,
            token_count,
            client_message_id,
            status,
            is_read,
        )

        result = await self.get_by_id(message_id)
        if not result:
            raise RuntimeError("Failed to retrieve created message")
        return result

    async def get_by_id(self, message_id: str) -> Message | None:
        """Get message by ID"""
        query = """
            SELECT
                id, conversation_id, role, content, message_type,
                media_urls, audio_url, audio_duration_seconds,
                token_count, client_message_id, created_at, metadata, status, is_read
            FROM messages
            WHERE id = $1
        """

        row = await db.fetchone(query, message_id)
        return self._row_to_message(row) if row else None

    async def get_by_client_id(self, conversation_id: str, client_message_id: str) -> Message | None:
        """Get message by client_message_id in a specific conversation"""
        query = """
            SELECT
                id, conversation_id, role, content, message_type,
                media_urls, audio_url, audio_duration_seconds,
                token_count, client_message_id, created_at, metadata
            FROM messages
            WHERE conversation_id = $1 AND client_message_id = $2
        """

        row = await db.fetchone(query, conversation_id, client_message_id)
        return self._row_to_message(row) if row else None

    async def get_assistant_reply(self, message_id: str) -> Message | None:
        """
        Get the assistant's reply for a specific user message.
        Assumes the reply is the first assistant message created after the user message
        in the same conversation.
        """
        msg = await self.get_by_id(message_id)
        if not msg:
            return None

        query = """
            SELECT
                id, conversation_id, role, content, message_type,
                media_urls, audio_url, audio_duration_seconds,
                token_count, client_message_id, created_at, metadata
            FROM messages
            WHERE conversation_id = $1
              AND role = 'assistant'
              AND created_at >= $2
              AND id != $3
            ORDER BY created_at ASC
            LIMIT 1
        """

        row = await db.fetchone(query, msg.conversation_id, msg.created_at, message_id)
        return self._row_to_message(row) if row else None

    async def list_by_conversation(
        self, conversation_id: str, limit: int = 50, offset: int = 0, order: str = "desc"
    ) -> list[Message]:
        """List messages in a conversation"""
        order_clause = "DESC" if order.lower() == "desc" else "ASC"

        query = (
            "SELECT "
            "id, conversation_id, role, content, message_type, "
            "media_urls, audio_url, audio_duration_seconds, "
            "token_count, client_message_id, created_at, metadata, status, is_read "
            "FROM messages "
            "WHERE conversation_id = $1 "
            "ORDER BY created_at " + order_clause + " "
            "LIMIT $2 OFFSET $3"
        )

        rows = await db.fetch(query, conversation_id, limit, offset)
        return [self._row_to_message(row) for row in rows]

    async def get_recent_for_context(self, conversation_id: str, limit: int = 10) -> list[Message]:
        """Get recent messages for AI context (ordered oldest to newest)"""
        query = """
            SELECT
                id, conversation_id, role, content, message_type,
                media_urls, audio_url, audio_duration_seconds,
                token_count, client_message_id, created_at, metadata, status, is_read
            FROM messages
            WHERE conversation_id = $1
            ORDER BY created_at DESC
            LIMIT $2
        """

        rows = await db.fetch(query, conversation_id, limit)
        return [self._row_to_message(row) for row in reversed(rows)]

    async def get_recent_for_conversations_batch(
        self, conversation_ids: list[str], limit_per_conv: int = 10
    ) -> dict[str, list[Message]]:
        """
        Get recent messages for multiple conversations efficiently.
        Returns a dictionary mapping conversation_id to list of messages (ordered oldest to newest).
        """
        if not conversation_ids:
            return {}

        # Dynamically build placeholders for the IN clause
        placeholders = ", ".join(f"${i + 1}" for i in range(len(conversation_ids)))

        # We need the limit as the last parameter
        limit_param_index = len(conversation_ids) + 1

        query = f"""
            WITH RankedMessages AS (
                SELECT
                    id, conversation_id, role, content, message_type,
                    media_urls, audio_url, audio_duration_seconds,
                    token_count, client_message_id, created_at, metadata, status, is_read,
                    ROW_NUMBER() OVER (
                        PARTITION BY conversation_id
                        ORDER BY created_at DESC
                    ) as rn
                FROM messages
                WHERE conversation_id IN ({placeholders})
            )
            SELECT *
            FROM RankedMessages
            WHERE rn <= ${limit_param_index}
            ORDER BY conversation_id, created_at ASC
        """

        # Combine arguments: conversation IDs + limit
        args = [*conversation_ids, limit_per_conv]

        rows = await db.fetch(query, *args)

        # Group by conversation_id
        result: dict[str, list[Message]] = {str(cid): [] for cid in conversation_ids}

        for row in rows:
            msg = self._row_to_message(row)
            if msg.conversation_id in result:
                result[msg.conversation_id].append(msg)

        return result

    async def mark_as_read(self, conversation_id: str) -> None:
        """Mark all messages in a conversation as read"""
        query = """
            UPDATE messages
            SET is_read = 1, status = 'read'
            WHERE conversation_id = $1 AND is_read = 0 AND role = 'assistant'
        """
        await db.execute(query, conversation_id)

    async def count_by_conversation(self, conversation_id: str) -> int:
        """Count messages in a conversation"""
        query = "SELECT COUNT(*) FROM messages WHERE conversation_id = $1"
        return int(await db.fetchval(query, conversation_id) or 0)

    async def count_all(self) -> int:
        """Count all messages"""
        query = "SELECT COUNT(*) FROM messages"
        return int(await db.fetchval(query) or 0)

    async def delete_by_conversation(self, conversation_id: str) -> int:
        """Delete all messages for a conversation and return count of deleted messages"""
        count_query = "SELECT COUNT(*) FROM messages WHERE conversation_id = $1"
        result = await db.fetchval(count_query, conversation_id)
        message_count = int(result) if result is not None else 0

        if message_count > 0:
            delete_query = "DELETE FROM messages WHERE conversation_id = $1"
            await db.execute(delete_query, conversation_id)

        return message_count

    def _row_to_message(self, row) -> Message:
        """Convert database row to Message model"""
        media_urls = row["media_urls"]
        if isinstance(media_urls, str):
            media_urls = json.loads(media_urls)
        elif not isinstance(media_urls, list):
            media_urls = []

        metadata = row["metadata"]
        if isinstance(metadata, str):
            metadata = json.loads(metadata)

        return Message(
            id=row["id"],
            conversation_id=row["conversation_id"],
            role=MessageRole(row["role"]),
            content=row["content"],
            message_type=MessageType(row["message_type"]),
            media_urls=media_urls,
            audio_url=row["audio_url"],
            audio_duration_seconds=row["audio_duration_seconds"],
            token_count=row["token_count"],
            client_message_id=row.get("client_message_id"),
            created_at=row["created_at"],
            status=row.get("status", "delivered"),
            is_read=bool(row.get("is_read", False)),
            metadata=metadata,
        )

File: src/db/repositories/message_repository.rs
use std::collections::HashMap;

#[cfg(not(feature = "staging"))]
use sqlx::PgPool;
#[cfg(feature = "staging")]
use sqlx::SqlitePool;

use uuid::Uuid;

#[cfg(feature = "staging")]
use super::parse_dt;

use crate::models::entities::{Message, MessageRole, MessageType};

// ── Staging: SQLite-only ──────────────────────────────────────────────────────

#[cfg(feature = "staging")]
pub struct MessageRepository {
    pool: SqlitePool,
}

#[cfg(feature = "staging")]
#[derive(sqlx::FromRow)]
struct MessageRow {
    id: String,
    conversation_id: String,
    role: String,
    content: Option<String>,
    message_type: String,
    media_urls: String,
    audio_url: Option<String>,
    audio_duration_seconds: Option<i32>,
    token_count: Option<i32>,
    client_message_id: Option<String>,
    created_at: String,
    metadata: String,
    status: Option<String>,
    is_read: Option<i32>,
}

#[cfg(feature = "staging")]
impl From<MessageRow> for Message {
    fn from(row: MessageRow) -> Self {
        Self {
            id: row.id,
            conversation_id: row.conversation_id,
            role: row.role.parse().unwrap_or(MessageRole::User),
            content: row.content,
            message_type: row.message_type.parse().unwrap_or(MessageType::Text),
            media_urls: serde_json::from_str(&row.media_urls).unwrap_or_default(),
            audio_url: row.audio_url,
            audio_duration_seconds: row.audio_duration_seconds,
            token_count: row.token_count,
            client_message_id: row.client_message_id,
            created_at: parse_dt(&row.created_at),
            metadata: serde_json::from_str(&row.metadata)
                .unwrap_or(serde_json::Value::Object(Default::default())),
            status: row.status.unwrap_or("delivered".to_string()),
            is_read: row.is_read.unwrap_or(0) != 0,
        }
    }
}

#[cfg(feature = "staging")]
const SELECT_COLS: &str = "id, conversation_id, role, content, message_type, media_urls, audio_url,
     audio_duration_seconds, token_count, client_message_id, created_at, metadata,
     status, is_read";

#[cfg(feature = "staging")]
impl MessageRepository {
    pub fn new(pool: SqlitePool) -> Self {
        Self { pool }
    }

    // ── Writes ────────────────────────────────────────────────────────────────

    #[allow(clippy::too_many_arguments)]
    pub async fn create(
        &self,
        conversation_id: &str,
        role: &MessageRole,
        content: Option<&str>,
        message_type: &MessageType,
        media_urls: &[String],
        audio_url: Option<&str>,
        audio_duration_seconds: Option<i32>,
        token_count: Option<i32>,
        client_message_id: Option<&str>,
    ) -> Result<Message, sqlx::Error> {
        let message_id = Uuid::new_v4().to_string();
        let media_urls_json = serde_json::to_string(media_urls).unwrap_or("[]".to_string());

        sqlx::query(
            "INSERT INTO messages (
                id, conversation_id, role, content, message_type,
                media_urls, audio_url, audio_duration_seconds, token_count,
                client_message_id, status, is_read
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
        )
        .bind(&message_id)
        .bind(conversation_id)
        .bind(role.as_ref())
        .bind(content)
        .bind(message_type.as_ref())
        .bind(&media_urls_json)
        .bind(audio_url)
        .bind(audio_duration_seconds)
        .bind(token_count)
        .bind(client_message_id)
        .bind("delivered")
        .bind(0)
        .execute(&self.pool)
        .await?;

        sqlx::query("UPDATE conversations SET updated_at = CURRENT_TIMESTAMP WHERE id = ?")
            .bind(conversation_id)
            .execute(&self.pool)
            .await?;

        self.get_by_id(&message_id)
            .await?
            .ok_or(sqlx::Error::RowNotFound)
    }

    pub async fn delete_by_conversation(&self, conversation_id: &str) -> Result<i64, sqlx::Error> {
        let count: (i64,) =
            sqlx::query_as("SELECT COUNT(*) FROM messages WHERE conversation_id = ?")
                .bind(conversation_id)
                .fetch_one(&self.pool)
                .await?;

        if count.0 > 0 {
            sqlx::query("DELETE FROM messages WHERE conversation_id = ?")
                .bind(conversation_id)
                .execute(&self.pool)
                .await?;
        }

        Ok(count.0)
    }

    pub async fn mark_as_read(&self, conversation_id: &str) -> Result<(), sqlx::Error> {
        sqlx::query(
            "UPDATE messages SET is_read = 1, status = 'read'
             WHERE conversation_id = ? AND is_read = 0 AND role = 'assistant'",
        )
        .bind(conversation_id)
        .execute(&self.pool)
        .await?;
        Ok(())
    }

    // ── Reads ─────────────────────────────────────────────────────────────────

    pub async fn get_by_id(&self, message_id: &str) -> Result<Option<Message>, sqlx::Error> {
        let row = sqlx::query_as::<_, MessageRow>(&format!(
            "SELECT {SELECT_COLS} FROM messages WHERE id = ?"
        ))
        .bind(message_id)
        .fetch_optional(&self.pool)
        .await?;
        Ok(row.map(Message::from))
    }

    pub async fn get_by_client_id(
        &self,
        conversation_id: &str,
        client_message_id: &str,
    ) -> Result<Option<Message>, sqlx::Error> {
        let row = sqlx::query_as::<_, MessageRow>(&format!(
            "SELECT {SELECT_COLS} FROM messages WHERE conversation_id = ? AND client_message_id = ?"
        ))
        .bind(conversation_id)
        .bind(client_message_id)
        .fetch_optional(&self.pool)
        .await?;
        Ok(row.map(Message::from))
    }

    pub async fn get_assistant_reply(
        &self,
        message_id: &str,
    ) -> Result<Option<Message>, sqlx::Error> {
        let msg = match self.get_by_id(message_id).await? {
            Some(m) => m,
            None => return Ok(None),
        };
        let row = sqlx::query_as::<_, MessageRow>(&format!(
            "SELECT {SELECT_COLS} FROM messages
             WHERE conversation_id = ? AND role = 'assistant'
               AND created_at >= ? AND id != ?
             ORDER BY created_at ASC LIMIT 1"
        ))
        .bind(&msg.conversation_id)
        .bind(msg.created_at.format("%Y-%m-%d %H:%M:%S").to_string())
        .bind(message_id)
        .fetch_optional(&self.pool)
        .await?;
        Ok(row.map(Message::from))
    }

    pub async fn list_by_conversation(
        &self,
        conversation_id: &str,
        limit: i64,
        offset: i64,
        order: &str,
    ) -> Result<Vec<Message>, sqlx::Error> {
        let order_clause = if order == "asc" { "ASC" } else { "DESC" };
        let sql = format!(
            "SELECT {SELECT_COLS} FROM messages
             WHERE conversation_id = ?
             ORDER BY created_at {order_clause}
             LIMIT ? OFFSET ?"
        );
        let rows = sqlx::query_as::<_, MessageRow>(&sql)
            .bind(conversation_id)
            .bind(limit)
            .bind(offset)
            .fetch_all(&self.pool)
            .await?;
        Ok(rows.into_iter().map(Message::from).collect())
    }

    pub async fn get_recent_for_context(
        &self,
        conversation_id: &str,
        limit: i64,
    ) -> Result<Vec<Message>, sqlx::Error> {
        let rows = sqlx::query_as::<_, MessageRow>(&format!(
            "SELECT {SELECT_COLS} FROM messages
             WHERE conversation_id = ?
             ORDER BY created_at DESC LIMIT ?"
        ))
        .bind(conversation_id)
        .bind(limit)
        .fetch_all(&self.pool)
        .await?;
        let mut messages: Vec<Message> = rows.into_iter().map(Message::from).collect();
        messages.reverse();
        Ok(messages)
    }

    pub async fn get_recent_for_conversations_batch(
        &self,
        conversation_ids: &[String],
        limit_per_conv: i64,
    ) -> Result<HashMap<String, Vec<Message>>, sqlx::Error> {
        if conversation_ids.is_empty() {
            return Ok(HashMap::new());
        }

        let placeholders: Vec<&str> = conversation_ids.iter().map(|_| "?").collect();
        let sql = format!(
            "WITH RankedMessages AS (
                SELECT {SELECT_COLS},
                       ROW_NUMBER() OVER (
                           PARTITION BY conversation_id ORDER BY created_at DESC
                       ) as rn
                FROM messages WHERE conversation_id IN ({})
            )
            SELECT {SELECT_COLS} FROM RankedMessages
            WHERE rn <= ? ORDER BY conversation_id, created_at ASC",
            placeholders.join(", ")
        );

        let mut query = sqlx::query_as::<_, MessageRow>(&sql);
        for id in conversation_ids {
            query = query.bind(id);
        }
        query = query.bind(limit_per_conv);

        let rows = query.fetch_all(&self.pool).await?;
        let mut result: HashMap<String, Vec<Message>> = conversation_ids
            .iter()
            .map(|id| (id.clone(), Vec::new()))
            .collect();
        for row in rows {
            let conv_id = row.conversation_id.clone();
            if let Some(messages) = result.get_mut(&conv_id) {
                messages.push(Message::from(row));
            }
        }
        Ok(result)
    }

    pub async fn count_by_conversation(&self, conversation_id: &str) -> Result<i64, sqlx::Error> {
        let count: (i64,) =
            sqlx::query_as("SELECT COUNT(*) FROM messages WHERE conversation_id = ?")
                .bind(conversation_id)
                .fetch_one(&self.pool)
                .await?;
        Ok(count.0)
    }

    pub async fn count_unread(&self, conversation_id: &str) -> Result<i64, sqlx::Error> {
        let count: (i64,) = sqlx::query_as(
            "SELECT COUNT(*) FROM messages WHERE conversation_id = ? AND is_read = 0 AND role = 'assistant'",
        )
        .bind(conversation_id)
        .fetch_one(&self.pool)
        .await?;
        Ok(count.0)
    }
}

// ── Non-staging: PostgreSQL-only ──────────────────────────────────────────────

#[cfg(not(feature = "staging"))]
pub struct MessageRepository {
    pg_pool: PgPool,
}

#[cfg(not(feature = "staging"))]
#[derive(sqlx::FromRow)]
struct PgMessageRow {
    id: String,
    conversation_id: String,
    role: String,
    content: Option<String>,
    message_type: String,
    media_urls: serde_json::Value,
    audio_url: Option<String>,
    audio_duration_seconds: Option<i32>,
    token_count: Option<i32>,
    client_message_id: Option<String>,
    created_at: chrono::NaiveDateTime,
    metadata: serde_json::Value,
    status: Option<String>,
    is_read: Option<bool>,
}

#[cfg(not(feature = "staging"))]
impl From<PgMessageRow> for Message {
    fn from(row: PgMessageRow) -> Self {
        Self {
            id: row.id,
            conversation_id: row.conversation_id,
            role: row.role.parse().unwrap_or(MessageRole::User),
            content: row.content,
            message_type: row.message_type.parse().unwrap_or(MessageType::Text),
            media_urls: serde_json::from_value(row.media_urls).unwrap_or_default(),
            audio_url: row.audio_url,
            audio_duration_seconds: row.audio_duration_seconds,
            token_count: row.token_count,
            client_message_id: row.client_message_id,
            created_at: row.created_at,
            metadata: row.metadata,
            status: row.status.unwrap_or("delivered".to_string()),
            is_read: row.is_read.unwrap_or(false),
        }
    }
}

#[cfg(not(feature = "staging"))]
const SELECT_COLS: &str = "id, conversation_id, role, content, message_type, media_urls, audio_url,
     audio_duration_seconds, token_count, client_message_id, created_at, metadata,
     status, is_read";

#[cfg(not(feature = "staging"))]
impl MessageRepository {
    pub fn new(pg_pool: PgPool) -> Self {
        Self { pg_pool }
    }

    // ── Writes ────────────────────────────────────────────────────────────────

    #[allow(clippy::too_many_arguments)]
    pub async fn create(
        &self,
        conversation_id: &str,
        role: &MessageRole,
        content: Option<&str>,
        message_type: &MessageType,
        media_urls: &[String],
        audio_url: Option<&str>,
        audio_duration_seconds: Option<i32>,
        token_count: Option<i32>,
        client_message_id: Option<&str>,
    ) -> Result<Message, sqlx::Error> {
        let message_id = Uuid::new_v4().to_string();
        let media_urls_json =
            serde_json::to_value(media_urls).unwrap_or(serde_json::Value::Array(vec![]));

        sqlx::query(
            "INSERT INTO messages (
                id, conversation_id, role, content, message_type,
                media_urls, audio_url, audio_duration_seconds, token_count,
                client_message_id, status, is_read
            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)",
        )
        .bind(&message_id)
        .bind(conversation_id)
        .bind(role.as_ref())
        .bind(content)
        .bind(message_type.as_ref())
        .bind(&media_urls_json)
        .bind(audio_url)
        .bind(audio_duration_seconds)
        .bind(token_count)
        .bind(client_message_id)
        .bind("delivered")
        .bind(false)
        .execute(&self.pg_pool)
        .await?;

        sqlx::query("UPDATE conversations SET updated_at = NOW() WHERE id = $1")
            .bind(conversation_id)
            .execute(&self.pg_pool)
            .await?;

        self.get_by_id(&message_id)
            .await?
            .ok_or(sqlx::Error::RowNotFound)
    }

    pub async fn delete_by_conversation(&self, conversation_id: &str) -> Result<i64, sqlx::Error> {
        let count: (i64,) =
            sqlx::query_as("SELECT COUNT(*) FROM messages WHERE conversation_id = $1")
                .bind(conversation_id)
                .fetch_one(&self.pg_pool)
                .await?;

        if count.0 > 0 {
            sqlx::query("DELETE FROM messages WHERE conversation_id = $1")
                .bind(conversation_id)
                .execute(&self.pg_pool)
                .await?;
        }

        Ok(count.0)
    }

    pub async fn mark_as_read(&self, conversation_id: &str) -> Result<(), sqlx::Error> {
        sqlx::query(
            "UPDATE messages SET is_read = TRUE, status = 'read'
             WHERE conversation_id = $1 AND is_read = FALSE AND role = 'assistant'",
        )
        .bind(conversation_id)
        .execute(&self.pg_pool)
        .await?;
        Ok(())
    }

    // ── Reads ─────────────────────────────────────────────────────────────────

    pub async fn get_by_id(&self, message_id: &str) -> Result<Option<Message>, sqlx::Error> {
        let row = sqlx::query_as::<_, PgMessageRow>(&format!(
            "SELECT {SELECT_COLS} FROM messages WHERE id = $1"
        ))
        .bind(message_id)
        .fetch_optional(&self.pg_pool)
        .await?;
        Ok(row.map(Message::from))
    }

    pub async fn get_by_client_id(
        &self,
        conversation_id: &str,
        client_message_id: &str,
    ) -> Result<Option<Message>, sqlx::Error> {
        let row = sqlx::query_as::<_, PgMessageRow>(&format!(
            "SELECT {SELECT_COLS} FROM messages
             WHERE conversation_id = $1 AND client_message_id = $2"
        ))
        .bind(conversation_id)
        .bind(client_message_id)
        .fetch_optional(&self.pg_pool)
        .await?;
        Ok(row.map(Message::from))
    }

    pub async fn get_assistant_reply(
        &self,
        message_id: &str,
    ) -> Result<Option<Message>, sqlx::Error> {
        let msg = match self.get_by_id(message_id).await? {
            Some(m) => m,
            None => return Ok(None),
        };
        let row = sqlx::query_as::<_, PgMessageRow>(&format!(
            "SELECT {SELECT_COLS} FROM messages
             WHERE conversation_id = $1 AND role = 'assistant'
               AND created_at >= $2 AND id != $3
             ORDER BY created_at ASC LIMIT 1"
        ))
        .bind(&msg.conversation_id)
        .bind(msg.created_at)
        .bind(message_id)
        .fetch_optional(&self.pg_pool)
        .await?;
        Ok(row.map(Message::from))
    }

    pub async fn list_by_conversation(
        &self,
        conversation_id: &str,
        limit: i64,
        offset: i64,
        order: &str,
    ) -> Result<Vec<Message>, sqlx::Error> {
        let order_clause = if order == "asc" { "ASC" } else { "DESC" };
        let sql = format!(
            "SELECT {SELECT_COLS} FROM messages
             WHERE conversation_id = $1
             ORDER BY created_at {order_clause}
             LIMIT $2 OFFSET $3"
        );
        let rows = sqlx::query_as::<_, PgMessageRow>(&sql)
            .bind(conversation_id)
            .bind(limit)
            .bind(offset)
            .fetch_all(&self.pg_pool)
            .await?;
        Ok(rows.into_iter().map(Message::from).collect())
    }

    pub async fn get_recent_for_context(
        &self,
        conversation_id: &str,
        limit: i64,
    ) -> Result<Vec<Message>, sqlx::Error> {
        let rows = sqlx::query_as::<_, PgMessageRow>(&format!(
            "SELECT {SELECT_COLS} FROM messages
             WHERE conversation_id = $1
             ORDER BY created_at DESC LIMIT $2"
        ))
        .bind(conversation_id)
        .bind(limit)
        .fetch_all(&self.pg_pool)
        .await?;
        let mut messages: Vec<Message> = rows.into_iter().map(Message::from).collect();
        messages.reverse();
        Ok(messages)
    }

    pub async fn get_recent_for_conversations_batch(
        &self,
        conversation_ids: &[String],
        limit_per_conv: i64,
    ) -> Result<HashMap<String, Vec<Message>>, sqlx::Error> {
        if conversation_ids.is_empty() {
            return Ok(HashMap::new());
        }

        let rows = sqlx::query_as::<_, PgMessageRow>(&format!(
            "WITH RankedMessages AS (
                SELECT {SELECT_COLS},
                       ROW_NUMBER() OVER (
                           PARTITION BY conversation_id ORDER BY created_at DESC
                       ) as rn
                FROM messages WHERE conversation_id = ANY($1)
            )
            SELECT {SELECT_COLS} FROM RankedMessages
            WHERE rn <= $2 ORDER BY conversation_id, created_at ASC"
        ))
        .bind(conversation_ids.to_vec())
        .bind(limit_per_conv)
        .fetch_all(&self.pg_pool)
        .await?;

        let mut result: HashMap<String, Vec<Message>> = conversation_ids
            .iter()
            .map(|id| (id.clone(), Vec::new()))
            .collect();
        for row in rows {
            let conv_id = row.conversation_id.clone();
            if let Some(messages) = result.get_mut(&conv_id) {
                messages.push(Message::from(row));
            }
        }
        Ok(result)
    }

    pub async fn count_by_conversation(&self, conversation_id: &str) -> Result<i64, sqlx::Error> {
        let count: (i64,) =
            sqlx::query_as("SELECT COUNT(*) FROM messages WHERE conversation_id = $1")
                .bind(conversation_id)
                .fetch_one(&self.pg_pool)
                .await?;
        Ok(count.0)
    }

    pub async fn count_unread(&self, conversation_id: &str) -> Result<i64, sqlx::Error> {
        let count: (i64,) = sqlx::query_as(
            "SELECT COUNT(*) FROM messages WHERE conversation_id = $1 AND is_read = FALSE AND role = 'assistant'",
        )
        .bind(conversation_id)
        .fetch_one(&self.pg_pool)
        .await?;
        Ok(count.0)
    }
}
